org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/GradleClasspathContainerInitializer.java

```diff
@@ -171,6 +172,12 @@ public IClasspathEntry apply(OmniEclipseSourceDirectory directory) {
                 // @formatter:on
             }
         }).toList();
+
+        return removeDuplicatesFromSourceFolders(sourceFolders);
+    }
+
+    private List<IClasspathEntry> removeDuplicatesFromSourceFolders(List<IClasspathEntry> sourceFolders) {
+        return ImmutableSet.copyOf(sourceFolders).asList();
```

Comment:
Does ImmutableSet preserve ordering ? normally classpath lists has an order thus if a set gets involved we'll need to make sure it guarantees order.